src/databricks/labs/ucx/source_code/notebook.py

```diff
@@ -0,0 +1,331 @@
+        pass  # not in scope
+
+
+class ScalaCell(Cell):
+
+    @property
+    def language(self):
+        return CellLanguage.SCALA
+
+    def is_runnable(self) -> bool:
+        return True  # TODO
+
+    def build_dependency_graph(self, parent: NotebookDependencyGraph):
+        pass  # TODO
+
+
+class SQLCell(Cell):
+
+    @property
+    def language(self):
+        return CellLanguage.SQL
+
+    def is_runnable(self) -> bool:
+        try:
+            statements = parse_sql(self._original_code)
+            return len(statements) > 0
+        except SQLParseError:
+            return False
+
+    def build_dependency_graph(self, parent: NotebookDependencyGraph):
+        pass  # not in scope
+
+
+class MarkdownCell(Cell):
+
+    @property
+    def language(self):
+        return CellLanguage.MARKDOWN
+
+    def is_runnable(self) -> bool:
+        return True  # TODO
+
+    def build_dependency_graph(self, parent: NotebookDependencyGraph):
+        pass  # not in scope
+
+
+class RunCell(Cell):
+
+    @property
+    def language(self):
+        return CellLanguage.RUN
+
+    def is_runnable(self) -> bool:
+        return True  # TODO
+
+    def build_dependency_graph(self, parent: NotebookDependencyGraph):
+        command = f'{LANGUAGE_PREFIX}{self.language.magic_name}'.strip()
+        lines = self._original_code.split('\n')
+        for line in lines:
+            start = line.index(command)
+            if start >= 0:
+                path = line[start + len(command) :].strip()
+                parent.register_dependency(path.strip('"'))
+                return
+        raise ValueError("Missing notebook path in %run command")
+
+
+class CellLanguage(Enum):
+    # long magic_names must come first to avoid shorter ones being matched
+    PYTHON = Language.PYTHON, 'python', '#', PythonCell
+    SCALA = Language.SCALA, 'scala', '//', ScalaCell
+    SQL = Language.SQL, 'sql', '--', SQLCell
+    RUN = None, 'run', None, RunCell
+    MARKDOWN = None, 'md', None, MarkdownCell
+    R = Language.R, 'r', '#', RCell
+
+    def __init__(self, *args):
+        super().__init__()
+        self._language = args[0]
+        self._magic_name = args[1]
+        self._comment_prefix = args[2]
+        self._new_cell = args[3]
+
+    @property
+    def language(self) -> Language:
+        return self._language
+
+    @property
+    def magic_name(self) -> str:
+        return self._magic_name
+
+    @property
+    def comment_prefix(self) -> str:
+        return self._comment_prefix
+
+    @classmethod
+    def of_language(cls, language: Language) -> CellLanguage:
+        return next((cl for cl in CellLanguage if cl.language == language))
+
+    @classmethod
+    def of_magic_name(cls, magic_name: str) -> CellLanguage | None:
+        return next((cl for cl in CellLanguage if magic_name.startswith(cl.magic_name)), None)
+
+    def read_cell_language(self, lines: list[str]) -> CellLanguage | None:
```

Comment:
```suggestion
    def of_lines(self, lines: list[str]) -> CellLanguage | None:
```

Reply:
not sure about this one, the `of_xxx` methods are static and selecting an Enum value from their property, this instance method has a very different contract

Reply:
this method is still static, no?